components/builder-api/src/server/resources/pkgs.rs

```diff
@@ -420,6 +423,142 @@ async fn download_package(req: HttpRequest,
+                                        let end_str = ent.get("period")?.get("end")?.as_str()?;
+                                        match NaiveDate::parse_from_str(end_str, "%Y-%m-%d") {
+                                            Ok(end_date) if end_date >= today => Some(end_date),
+                                            _ => None,
+                                        }
+                                    });
+
+                                    match refreshed_expiration {
+                                        Some(new_expiration) => {
+                                            if new_expiration >= today {
+                                                let update =
+                                                    NewLicenseKey { account_id:
+                                                                        account_id.try_into()
+                                                                                  .unwrap(),
+                                                                    license_key:
+                                                                        &license.license_key,
+                                                                    expiration_date: new_expiration, };
+
+                                                if let Err(err) = LicenseKey::create(&update, &conn)
+                                                {
+                                                    debug!("Failed to update license in DB: {}",
+                                                           err);
+                                                    return HttpResponse::InternalServerError().body("License update failed.");
+                                                }
+                                            }
+                                        }
+                                        None => {
+                                            return HttpResponse::Forbidden().body("Your license \
+                                                                                   has expired.");
+                                        }
+                                    }
+                                }
+                            }
+                            Ok(None) => {
+                                return HttpResponse::Forbidden().body("No valid license key \
+                                                                       found.");
+                            }
+                            Err(err) => {
+                                debug!("License DB error: {}", err);
+                                return HttpResponse::InternalServerError().body("License validation error.");
+                            }
+                        }
+                    }
+                    None => {
+                        return HttpResponse::Unauthorized().body("Authentication required.");
+                    }
+                }
+            }
+
```

Comment:
we now have two locations where we fetch the license and check the expiration: here and in the license input. Can we consolidate the logic?